services/mfa/mfa_test.go

```diff
@@ -27,26 +28,45 @@ func TestGenerateSecret(t *testing.T) {
 	config.SetDefaults()
 	config.ServiceSettings.EnableMultifactorAuthentication = model.NewBool(true)
 	configService := testutils.StaticConfigService{Cfg: &config}
-	storeMock := mocks.Store{}
-	userStoreMock := mocks.UserStore{}
-	userStoreMock.On("UpdateMfaSecret", user.Id, mock.AnythingOfType("string")).Return(func(userId string, secret string) *model.AppError {
-		return nil
-	})
-	storeMock.On("User").Return(&userStoreMock)
 
-	mfa := Mfa{configService, &storeMock}
+	t.Run("fail on disabled mfa", func(t *testing.T) {
+		wrongConfig := model.Config{}
+		wrongConfig.SetDefaults()
+		wrongConfig.ServiceSettings.EnableMultifactorAuthentication = model.NewBool(false)
+		wrongConfigService := testutils.StaticConfigService{Cfg: &wrongConfig}
+		mfa := New(wrongConfigService, nil)
+		_, _, err := mfa.GenerateSecret(user)
+		require.NotNil(t, err)
+	})
 
-	secret, img, err := mfa.GenerateSecret(user)
-	require.Nil(t, err)
+	t.Run("fail on store action fail", func(t *testing.T) {
+		storeMock := mocks.Store{}
+		userStoreMock := mocks.UserStore{}
+		userStoreMock.On("UpdateMfaSecret", user.Id, mock.AnythingOfType("string")).Return(func(userId string, secret string) *model.AppError {
+			return model.NewAppError("GenerateQrCode", "mfa.generate_qr_code.save_secret.app_error", nil, "", http.StatusInternalServerError)
+		})
+		storeMock.On("User").Return(&userStoreMock)
 
-	assert.Len(t, secret, 32)
+		mfa := New(configService, &storeMock)
+		_, _, err := mfa.GenerateSecret(user)
+		require.NotNil(t, err)
+	})
 
-	require.NotEmpty(t, img, "no image set")
+	t.Run("Successful generate secret", func(t *testing.T) {
```

Comment:
Is it also possible to test for `qr.Encode(authLink, qr.H)` by passing some garbage `SiteURL` ?

Reply:
It is not that easy I think, I tried but the function `Encode` is pretty robust by itself.